### PR TITLE
Internationalization

### PR DESCRIPTION
**Pull request by tlrobinson**
I've extracted the internationalization work I've done into it's own PR here, and added locale negotiation based on what the user's browser sends in the `Accept-Language` header.

Basically, here's the workflow:

1. Tag strings in the frontend using `t` and `jt` ES6 template literals (see more details in https://c-3po.js.org/):

```javascript
const someString = t`Hello ${name}!`;
const someJSX = <div>{ jt`Hello ${name}` }</div>
```

and in the backend using `trs` and related macros (see more details in https://github.com/puppetlabs/clj-i18n):

```clojure
(trs "Hello {0}!" name)
```

2. When you have added/edited tagged strings in the code, run `./bin/i18n/update-translations` to update the base `locales/metabase.pot` template and each existing `locales/LOCALE.po`
3. To add a new translaction run `./bin/i18n/update-translation LOCALE`
4. Edit translation in `locales/LOCALE.po`
5. Run `./bin/i18n/build-translation-resources` to compile translations for frontend and backend
6. Restart or rebuild Metabase

To try it out, change your browser's language (e.x. chrome://settings/?search=language) to one of the locales to see it working. Run metabase with the `JAVA_TOOL_OPTIONS=-Duser.language=LOCALE` environment variable set to set the locale on the backend, e.x. for pulses and emails (eventually we'll also add a setting in the app)

### CODE REVIEW

File: locales/metabase.pot [hunk 1]
```diff
@@ -0,0 +1,138 @@
+# #-#-#-#-#  metabase-backend.pot (metabase)  #-#-#-#-#
```
Comment (senior) [inline locales/metabase.pot]: We'll need this initially, but I think we'll want to have something automated in place to generate these files and check them in. If not, pretty much every PR will conflict on this file. Any addition or removal of a line above/below a string will cause the line number to change an thus a change to this file.

File: src/metabase/util.clj [hunk 2]
```diff
@@ -29,7 +30,7 @@
 ;; This is the very first log message that will get printed.
 ;; It's here because this is one of the very first namespaces that gets loaded, and the first that has access to the logger
 ;; It shows up a solid 10-15 seconds before the "Starting Metabase in STANDALONE mode" message because so many other namespaces need to get loaded
-(log/info "Loading Metabase...")
+(log/info (trs "Loading {0}..." (trs "Metabase")))
```
Comment (senior) [inline src/metabase/util.clj]: You don't need the second `trs` here

File: src/metabase/routes.clj [hunk 3]
```diff
@@ -31,20 +33,31 @@
 (defn- load-template [path variables]
   (stencil/render-string (load-file-at-path path) variables))
 
+(defn- load-localization []
+  (log/info *locale*)
```
Comment (senior) [inline src/metabase/routes.clj]: We should have some metadata around this log messge. "Starting up with locale:... " or something similar

File: src/metabase/routes.clj [hunk 4]
```diff
@@ -31,20 +33,31 @@
...
+  (if (and *locale* (not= (str *locale*) "en"))
+    (try
+      (load-file-at-path (str "frontend_client/app/locales/" *locale* ".json"))
+    (catch Throwable e
+      (log/warn (str "Locale " *locale* " not found."))
+      "null"))
```
Comment (senior) [inline src/metabase/routes.clj]: Why return the string null?

File: src/metabase/core.clj [hunk 5]
```diff
@@ -133,6 +145,10 @@
     ;; start the metabot thread
     (metabot/start-metabot!))
 
+  (sync-locale)
+  (future (while true (do (Thread/sleep 1000) (sync-locale))))
```
Comment (senior) [inline src/metabase/core.clj]: What's the goal here?

File: src/metabase/api/setup.clj [hunk 6]
```diff
@@ -115,9 +116,9 @@
         num-tables         (db/count 'Table)
         num-cards          (db/count 'Card)
         num-users          (db/count 'User)]
-    [{:title       "Add a database"
-      :group       "Get connected"
-      :description "Connect to your data so your whole team can start to explore."
+    [{:title       (trs "Add a database")
```
Comment (senior) [inline src/metabase/api/setup.clj]: It looks like this is getting presented to the user. If that's the case, you want `tru` not `trs`. `trs` will return text in the system's local, not the (browser) user's locale.

File: src/metabase/routes.clj [hunk 7]
```diff
@@ -31,20 +33,31 @@
...
+                        :base_href         (escape-script (json/generate-string (base-href)))
+                        :embed_code        (when embeddable? (embed/head uri))})
         (load-file-at-path "frontend_client/init.html"))
       resp/response
       (resp/content-type "text/html; charset=utf-8")))
 
-(def ^:private index  (partial entrypoint "index"  (not :embeddable)))
-(def ^:private public (partial entrypoint "public" :embeddable))
-(def ^:private embed  (partial entrypoint "embed"  :embeddable))
+(def ^:private index  (locale-negotiator (partial entrypoint "index"  (not :embeddable))))
```
Comment (senior) [inline src/metabase/routes.clj]: It seems like we'd want to apply the local negotiator to all routes, not just index/public/embed. Is there a reason to limit it?

File: src/metabase/util.clj [hunk 8]
```diff
@@ -16,6 +16,7 @@
             [clojure.tools.namespace.find :as ns-find]
             colorize.core ; this needs to be loaded for `format-color`
             [metabase.config :as config]
+            [puppetlabs.i18n.core :as i18n :refer [trs trsn tru trun]]
```
Comment (senior) [inline src/metabase/util.clj]: My guess is this will fail our linters

File: src/metabase/routes.clj [hunk 9]
```diff
@@ -31,20 +33,31 @@
 (defn- load-template [path variables]
   (stencil/render-string (load-file-at-path path) variables))
 
+(defn- load-localization []
+  (log/info *locale*)
+  (if (and *locale* (not= (str *locale*) "en"))
+    (try
+      (load-file-at-path (str "frontend_client/app/locales/" *locale* ".json"))
+    (catch Throwable e
+      (log/warn (str "Locale " *locale* " not found."))
```
Comment (senior) [inline src/metabase/routes.clj]: You want this to be `trs`. We're logging that the user's locale couldn't be found, not the system's local.

File: src/metabase/core.clj [hunk 10]
```diff
@@ -31,7 +34,16 @@
...
+  (:import java.util.Locale
+           org.eclipse.jetty.server.Server))
+
+;; HACK: replace this with setting event listener
+(defn sync-locale []
+  (let [new-locale     (or (not-empty (setting/get :site-locale)) "en")
+        current-locale (.toLanguageTag (Locale/getDefault))]
+    (log/info (trs "Hello World!"))
+    (if (not= current-locale new-locale)
+      (Locale/setDefault (Locale/forLanguageTag new-locale)))))
```
Comment (senior) [inline src/metabase/core.clj]: I think you'd usually see a `when-not` here, or `(when (not...`

File: src/metabase/core.clj [hunk 11]
```diff
@@ -31,7 +34,16 @@
              [params :refer [wrap-params]]
              [session :refer [wrap-session]]]
             [toucan.db :as db])
-  (:import org.eclipse.jetty.server.Server))
+  (:import java.util.Locale
+           org.eclipse.jetty.server.Server))
+
+;; HACK: replace this with setting event listener
+(defn sync-locale []
+  (let [new-locale     (or (not-empty (setting/get :site-locale)) "en")
```
Comment (senior) [inline src/metabase/core.clj]: I think since you set a default, if there's no value it will already be "en"

**Review by senior — DISMISSED** [review]

File: src/metabase/util.clj (hunk 2 above)
Comment (tlrobinson) [reply src/metabase/util.clj]: Are you sure? I want "Metabase" itself to be a translated string.

File: src/metabase/routes.clj (hunk 3 above)
Comment (tlrobinson) [reply src/metabase/routes.clj]: Ah yeah that was added for debugging purposes, but yes, might be good to make it a real log.

File: src/metabase/routes.clj (hunk 4 above)
Comment (tlrobinson) [reply src/metabase/routes.clj]: This function returns serialized JSON.

File: src/metabase/api/setup.clj (hunk 6 above)
Comment (tlrobinson) [reply src/metabase/api/setup.clj]: 👍

File: src/metabase/core.clj (hunk 5 above)
Comment (tlrobinson) [reply src/metabase/core.clj]: This is a bit of a hack to keep the system locale in sync with the setting. See https://github.com/metabase/metabase/pull/5803/files#diff-a61f704fe001e6dad2115909432f583dR40 I tried adding an event listener like we have for metabot (for example) but couldn't get it working for some reason.

File: src/metabase/core.clj (hunk 10 above)
Comment (tlrobinson) [reply src/metabase/core.clj]: 👍

File: src/metabase/core.clj (hunk 11 above)
Comment (tlrobinson) [reply src/metabase/core.clj]: Hmm you're right. For some reason I was getting null at some point. Maybe I accidentally set the value to null.

File: src/metabase/util.clj (hunk 2 above)
Comment (salsakran) [reply src/metabase/util.clj]: I don't think it should be a translated string in this PR?

File: src/metabase/util.clj (hunk 8 above)
Comment (tlrobinson) [reply src/metabase/util.clj]: Fixed

File: src/metabase/core.clj (hunk 5 above)
Comment (tlrobinson) [reply src/metabase/core.clj]: @camsaul I took another look at this and it looks like we only publish the `settings-update` event for `set-many!` but not `set!` etc, as mentioned here https://github.com/metabase/metabase/blob/master/src/metabase/models/setting.clj#L381

I feel like it would actually be useful for that event to fire for all the other `set*!` functions too. It doesn't seem like there would be another easy way to be notified when the locale setting changes (aside from polling like I've done so far). If I followed the suggestion in the TODO linked above, I'd have to add a custom endpoint and a special case in the frontend to use that endpoint for the `site-locale` setting (or a special case in the `PUT /api/setting/:key` endpoint)

File: src/metabase/core.clj (hunk 11 above)
Comment (tlrobinson) [reply src/metabase/core.clj]: Fixed

File: src/metabase/routes.clj (hunk 9 above)
Comment (tlrobinson) [reply src/metabase/routes.clj]: Fixed

File: src/metabase/util.clj (hunk 2 above)
Comment (tlrobinson) [reply src/metabase/util.clj]: Fixed

File: locales/metabase.pot (hunk 1 above)
Comment (tlrobinson) [reply locales/metabase.pot]: So the idea is every time `master` is updated this automated thing would run `./bin/i18n/update-translations` and commit the changes?

Comment (camsaul) [reply locales/metabase.pot]: how would this script get ran, as a Git commit hook?

Comment (senior) [reply locales/metabase.pot]: @tlrobinson Is that script where the .po* files are being generated? Taking a look at the script in the PR that's what it looks like it's doing. The important point there is not relying on humans to remember to run that script and to get it out of PRs so that it doesn't cause merge conflicts.

@camsaul Right, I think we want a push hook that just listens on master described [here](https://github.com/metabase/metabase/issues/5893)

Comment (camsaul) [reply locales/metabase.pot]: That would totally bust development for people on Windows (including myself once or twice a week and several contributors) 😞

Comment (senior) [reply locales/metabase.pot]: Not sure I understand, what would bust windows developers?

Comment (camsaul) [reply locales/metabase.pot]: running bash scripts as part of git hooks

File: locales/de.po [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+"Content-Transfer-Encoding: 8bit\n"
+"#-#-#-#-#  metabase-frontend.pot  #-#-#-#-#\n"
+"Plural-Forms: nplurals=2; plural=(n != 1);\n"
+"#-#-#-#-#  metabase-backend.pot (metabase)  #-#-#-#-#\n"
+
+#: frontend/src/metabase/home/components/Activity.jsx:131
+msgid "Hello World!"
+msgstr "Hallo Welt!"
+
+#: frontend/src/metabase/home/components/Activity.jsx:132
```
Comment (camsaul) [inline locales/de.po]: does it make sense to include all these line numbers? They'll surely get out of date pretty quickly

File: src/metabase/api/setup.clj [hunk 13]
```diff
@@ -148,9 +149,9 @@
       :link        "/admin/datamodel/database"
       :completed   has-hidden-tables?
       :triggered   (>= num-tables 20)}
-     {:title       "Organize questions"
-      :group       "Curate your data"
-      :description "Have a lot of saved questions in Metabase? Create collections to help manage them and add context."
+     {:title       (tru "Organize questions")
+      :group       (tru "Curate your data")
+      :description (tru "Have a lot of saved questions in {0}? Create collections to help manage them and add context." (tru "Metabase"))
```
Comment (camsaul) [inline src/metabase/api/setup.clj]: can't someone translate 

```
Have a lot of saved questions in Metabase? Create collections to help manage them and add context.
```

as-is?

File: src/metabase/core.clj (hunk 5 above)
Comment (camsaul) [reply src/metabase/core.clj]: you don't need to do any of that stuff. You can give Settings customer magic setter functions. Take a look at `site-url` in `public_settings.clj` for an example. 

So here just give your `site-locale` setting a custom setter that calls that `sync-locale` function.

Comment (camsaul) [reply src/metabase/core.clj]: btw doing something like this is a bad idea because

1.  it's going to start calling `sync-locale` as soon as the namespace is evaluated, which means (if I understand what I think it's supposed to be doing) it will possibly be trying to use Settings before the DB is initialized (race condition). If the DB isn't ready yet it will throw an Exception that you won't end up seeing because it's on a background thread.
2. When that Exception gets thrown (or any other exception is raised by `sync-locale`)  the loop will break and `sync-locale` won't be called any additional times.

At any rate you don't need to schedule a recurring task here, but if you did, we have Quartz and the `metabase.task` stuff for that instead of reïnventing the wheel

File: src/metabase/core.clj [hunk 14]
```diff
@@ -133,6 +145,10 @@
     ;; start the metabot thread
     (metabot/start-metabot!))
 
+  (sync-locale)
+  (future (while true (do (Thread/sleep 1000) (sync-locale))))
+  ; (java.util.Locale/setDefault (new java.util.Locale "de"))
```
Comment (camsaul) [inline src/metabase/core.clj]: do we need this?

File: src/metabase/api/setup.clj (hunk 13 above)
Comment (camsaul) [reply src/metabase/api/setup.clj]: e.g. you have `Loading Metabase...` later on without making `Metabase` a param

File: src/metabase/core.clj [hunk 15]
```diff
@@ -50,6 +61,7 @@
       mb-middleware/wrap-api-key         ; looks for a Metabase API Key on the request and assocs as :metabase-api-key
       mb-middleware/wrap-session-id      ; looks for a Metabase Session ID and assoc as :metabase-session-id
       mb-middleware/maybe-set-site-url   ; set the value of `site-url` if it hasn't been set yet
+      locale-negotiator                  ; Binds *locale* for i18n
```
Comment (camsaul) [inline src/metabase/core.clj]: Does this just look at the `Accept-Language` header?

File: src/metabase/public_settings.clj [hunk 16]
```diff
@@ -143,6 +150,7 @@
    :google_auth_client_id (setting/get :google-auth-client-id)
    :has_sample_dataset    (db/exists? 'Database, :is_sample true)
    :ldap_configured       ((resolve 'metabase.integrations.ldap/ldap-configured?))
+   :available_locales     (vec (map (fn [locale] [locale (.getDisplayName (Locale/forLanguageTag locale))]) (available-locales)))
```
Comment (camsaul) [inline src/metabase/public_settings.clj]: `mapv` = `map` + `vec`. Although you shouldn't need that here because when the whole thing gets written to JSON it won't make a different whether it was a lazy seq or not.

The only reason you'd really need to realize it right away (using `vec` to force realization) would be if lazily evaluating it wouldn't work for some reason. For example if `available-locales` was using a DB cursor that would be closed by the time the lazy sequence was finally realized (when the JSON was being written). 

TL;DR You can get away with just suing `map` here

File: src/metabase/routes.clj (hunk 9 above)
Comment (camsaul) [reply src/metabase/routes.clj]: @senior `(trs *locale*)` does something differently than `(str *locale*)`?

File: locales/metabase.pot (hunk 1 above)
Comment (senior) [reply locales/metabase.pot]: I see where the misunderstanding is. It's a hook on the central repo, not a local git repo hook, see my link above. When a commit is pushed (and on master) we would have a process running on a separate server that would get POSTed to. That would cause the script to run on that server, check in the changes. 

The intent is to keep these kinds of things away from developers day to day operations. Since developers aren't messing with it, we won't have merge conflicts (or windows compatiblity issues).

Comment (camsaul) [reply locales/metabase.pot]: oh ok. Makes sense 👍

File: locales/de.po (hunk 12 above)
Comment (senior) [reply locales/de.po]: That's why in my opinion it's important to get these updates happening out of our normal dev flow. AFAIK having the line numbers is required for the translation software we're using. Any change to a file that changes the line number of a translation, needs to have this updated. Because of that, there's the potential for a lot of churn on this file. Also since it's a separate thing thats needs to be ran, we're pretty likely to forget it.

File: src/metabase/routes.clj (hunk 9 above)
Comment (tlrobinson) [reply src/metabase/routes.clj]: @camsaul No, `trs` just calls `(system-locale)` (https://github.com/puppetlabs/clj-i18n/blob/master/src/puppetlabs/i18n/core.clj#L279) and `tru` calls `(user-locale)` (https://github.com/puppetlabs/clj-i18n/blob/master/src/puppetlabs/i18n/core.clj#L269)

File: src/metabase/core.clj (hunk 5 above)
Comment (tlrobinson) [reply src/metabase/core.clj]: Ok, I'll use a custom setting setter.

The `sync-locale` thing was just a temporary thing to get it working.

File: src/metabase/api/setup.clj (hunk 13 above)
Comment (tlrobinson) [reply src/metabase/api/setup.clj]: `Metabase` will need to be a parameter for whitelabeling where you can change it to something else.

File: src/metabase/core.clj (hunk 15 above)
Comment (tlrobinson) [reply src/metabase/core.clj]: Yes https://github.com/puppetlabs/clj-i18n/blob/master/src/puppetlabs/i18n/core.clj#L342

File: src/metabase/routes.clj (hunk 9 above)
Comment (camsaul) [reply src/metabase/routes.clj]: 👍

File: src/metabase/core.clj (hunk 14 above)
Comment (tlrobinson) [reply src/metabase/core.clj]: Fixed

File: src/metabase/public_settings.clj (hunk 16 above)
Comment (tlrobinson) [reply src/metabase/public_settings.clj]: Done

File: src/metabase/core.clj (hunk 5 above)
Comment (tlrobinson) [reply src/metabase/core.clj]: Done.
